internal/backend/discussion_access.go

```diff
@@ -0,0 +1,169 @@
+package backend
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/nedrocks/delphisbe/internal/datastore"
+
+	"github.com/nedrocks/delphisbe/graph/model"
+	"github.com/sirupsen/logrus"
+	"go.uber.org/multierr"
+)
+
+func (d *delphisBackend) GetDiscussionAccessByUserID(ctx context.Context, userID string) ([]*model.Discussion, error) {
+	var discussions []*model.Discussion
+
+	// Get public discussions
+	publicIter := d.db.GetPublicDiscussions(ctx)
+	publicDiscussions, err := d.iterToDiscussions(ctx, publicIter)
+	if err != nil {
+		logrus.WithError(err).Error("failed to get public discussions")
+		return nil, err
+	}
+
+	// Get discussions the user has access to by flair
+	flairDiscIter := d.db.GetDiscussionsForFlairTemplateByUserID(ctx, userID)
+	flairDiscussions, err := d.iterToDiscussions(ctx, flairDiscIter)
+	if err != nil {
+		logrus.WithError(err).Error("failed to get flair discussions")
+		return nil, err
+	}
+
+	// Get discussions the user was invited to
+	userDiscIter := d.db.GetDiscussionsForUserAccessByUserID(ctx, userID)
+	userDiscussions, err := d.iterToDiscussions(ctx, userDiscIter)
+	if err != nil {
+		logrus.WithError(err).Error("failed to get user access discussions")
+		return nil, err
+	}
+
+	discussions = append(publicDiscussions, flairDiscussions...)
+	discussions = append(discussions, userDiscussions...)
+
+	return discussions, nil
+}
+
+func (d *delphisBackend) GetDiscussionFlairTemplateAccessByDiscussionID(ctx context.Context, discussionID string) ([]*model.FlairTemplate, error) {
+	iter := d.db.GetDiscussionFlairTemplatesAccessByDiscussionID(ctx, discussionID)
+	return d.iterToFlairTemplates(ctx, iter)
+}
+
+func (d *delphisBackend) PutDiscussionFlairTemplatesAccess(ctx context.Context, discussionID string, flairTemplateIDs []string) ([]*model.DiscussionFlairTemplateAccess, error) {
+	if len(flairTemplateIDs) == 0 {
+		return nil, fmt.Errorf("no flair template IDs to add to discussion")
+	}
+
+	var addedTemplates []*model.DiscussionFlairTemplateAccess
+
+	tx, err := d.db.BeginTx(ctx)
+	if err != nil {
+		logrus.WithError(err).Error("failed to begin tx")
+		return nil, err
+	}
+
+	for _, id := range flairTemplateIDs {
+		resp, err := d.db.UpsertDiscussionFlairTemplatesAccess(ctx, tx, discussionID, id)
```

Comment:
We need to ensure the moderator has the flair they want to add. They can't add flair they don't have (right now anyway).